lib/saluki-components/src/transforms/dogstatsd_mapper/mod.rs

```diff
@@ -0,0 +1,254 @@
+                        "profile: {}, mapping num {}: name is required",
+                        config_profile.name,
+                        i
+                    ));
+                }
+                if mapping.name == "" {
+                    return Err(generic_error!(
+                        "profile: {}, mapping num {}: match is required",
+                        config_profile.name,
+                        i
+                    ));
+                }
+                let regex = build_regex(&mapping.metric_match, &match_type)?;
+                profile.mappings.push(MetricMapping {
+                    name: mapping.name.clone(),
+                    tags: mapping.tags.clone(),
+                    regex,
+                });
+            }
+            profiles.push(profile);
+        }
+
+        Ok(MetricMapper { profiles })
+    }
+}
+
+impl DogstatsDMapperConfiguration {
+    /// Creates a new `DogstatsDMapperConfiguration` from the given configuration.
+    pub fn from_configuration(config: &GenericConfiguration) -> Result<Self, GenericError> {
+        let mut config: Self = config.as_typed()?;
+        config.context_string_interner_bytes = ByteSize::kib(512);
+        Ok(config)
+    }
+}
+
+#[async_trait]
+impl SynchronousTransformBuilder for DogstatsDMapperConfiguration {
+    async fn build(&self) -> Result<Box<dyn SynchronousTransform + Send>, GenericError> {
+        let metric_mapper = self.dogstatsd_mapper_profiles.build()?;
+        Ok(Box::new(DogstatsDMapper {
+            context_string_interner_bytes: self.context_string_interner_bytes,
+            metric_mapper,
+        }))
+    }
+}
+
+impl MemoryBounds for DogstatsDMapperConfiguration {
+    fn specify_bounds(&self, builder: &mut MemoryBoundsBuilder) {
+        builder
+            .minimum()
+            // Capture the size of the heap allocation when the component is built.
+            .with_single_value::<DogstatsDMapper>("component struct")
+            // We also allocate the backing storage for the string interner up front, which is used by our context
+            // resolver.
+            .with_fixed_amount("string interner", self.context_string_interner_bytes.as_u64() as usize);
+    }
+}
+
+pub struct DogstatsDMapper {
+    context_string_interner_bytes: ByteSize,
+    metric_mapper: MetricMapper,
+}
+
+impl DogstatsDMapper {}
+
+impl SynchronousTransform for DogstatsDMapper {
+    fn transform_buffer(&self, event_buffer: &mut FixedSizeEventBuffer) {
+        let context_string_interner_size = NonZeroUsize::new(self.context_string_interner_bytes.as_u64() as usize)
+            .ok_or_else(|| generic_error!("context_string_interner_size must be greater than 0"))
+            .unwrap();
+        let mut context_resolver = ContextResolverBuilder::from_name("dogstatsd_mapper")
+            .expect("resolver name is not empty")
+            .with_interner_capacity_bytes(context_string_interner_size)
+            .build();
```

Comment:
You need to build the resolver when we build the component itself, and then stick it in `DogstatsDMapper` and use it from there... otherwise we're never really interning. :P

Reply:
Might even be easiest just to pass it directly into `MetricMapper` and use it internally so that the higher-level transform code doesn't need to futz around with it.

Reply:
I had that originally but I think I ran into the issue of `transform_buffer` useing `&self` and to call `context_resolver.resolve(...)` I needed a `&mut self` on the `map` method.

Reply:
Hmm, yes, you're right.

I don't know why I made `SynchronousTransform::transform_buffer` take `&self` instead of `&mut self`.

I'm gonna go fix that right now in a small PR.